_posts/2018-02-05-chaining-http-requests-in-elm.md

```diff
@@ -0,0 +1,340 @@
+
+```haskell
+module Main exposing (main)
+
+import Html exposing (..)
+import Http
+import Json.Decode as Decode
+import Task exposing (Task)
+
+
+type alias Model =
+    { name : Maybe String
+    , error : String
+    }
+
+type Msg
+    = Name (Result String String)
+
+eventsRequest : Http.Request (List String)
+eventsRequest =
+    Http.get "https://api.github.com/events"
+        (Decode.list (Decode.at [ "actor", "login" ] Decode.string))
+
+nameRequest : String -> Http.Request String
+nameRequest login =
+    Http.get ("https://api.github.com/users/" ++ login)
+        (Decode.at [ "name" ]
+            (Decode.oneOf
+                [ Decode.string
+                , Decode.null "unspecified"
+                ]
+            )
+        )
+
+toHttpTask : Http.Request a -> Task String a
+toHttpTask request =
+    request
+        |> Http.toTask
+        |> Task.mapError toString
+
+fetchEvents : Task String (List String)
+fetchEvents =
+    toHttpTask eventsRequest
+
+fetchName : String -> Task String String
+fetchName login =
+    toHttpTask (nameRequest login)
+
+pickFirst : List String -> Task String String
+pickFirst events =
+    case List.head events of
+        Just event ->
+            Task.succeed event
+
+        Nothing ->
+            Task.fail "No events."
+
+init : ( Model, Cmd Msg )
+init =
+    { name = Nothing, error = "" }
+        ! [ fetchEvents
+                |> Task.andThen pickFirst
+                |> Task.andThen fetchName
+                |> Task.attempt Name
+          ]
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        Name (Ok name) ->
+            { model | name = Just name } ! []
+
+        Name (Err error) ->
+            { model | error = error } ! []
+
+view : Model -> Html Msg
+view model =
+    div []
+        [ if model.error /= "" then
+            div []
+                [ h4 [] [ text "Error encountered" ]
+                , pre [] [ text model.error ]
+                ]
+          else
+            text ""
+        , p [] [ text <| Maybe.withDefault "Fetching..." model.name ]
+        ]
+
+main =
+    Html.program
+        { init = init
+        , update = update
+        , subscriptions = always Sub.none
+        , view = view
+        }
+```
+
+That's for sure more code than with the JavaScript example, but don't forget that the Elm version renders HTML, not just logs in the console, and that the JavaScript code could be refactored to look a lot like the Elm version.
+
+As always, an [Ellie](https://ellie-app.com/7Q9svdqRGa1/3) is publicly available so you can play around with the code.
+
+If anything, the main takeaways from this post are these ones:
+
+- Learn [how to read Elm function signatures],
+- Learn how [Generic Types] work,
+- Keep trying understanding the deepest bits when facing something you don't fully understand,
```

Comment:
`s/trying understanding/trying to understand/`

Also, I'm not sure about this advice... at least, I'm sure I don't follow when I'm learning some new piece of technology: I'd rather not go down each and every rabbit hole before at least having some global understanding of the "thing".